metricbeat/mb/registry.go

```diff
@@ -300,11 +300,41 @@ func (r *Register) Modules() []string {
 	r.lock.RLock()
 	defer r.lock.RUnlock()
 
+	var dups = map[string]bool{}
+
+	// For the sake of compatibility, grab modules the old way as well, right from the modules map
 	modules := make([]string, 0, len(r.modules))
 	for module := range r.modules {
+		dups[module] = true
 		modules = append(modules, module)
 	}
 
+	// List also modules from secondary sources
+	if source := r.secondarySource; source != nil {
+		sourceModules, err := source.Modules()
+		if err != nil {
+			logp.L().Errorf("failed to get modules from secondary source: %s", err)
+		} else {
+			for _, module := range sourceModules {
+				if _, ok := dups[module]; ok {
+					continue
+				}
+				dups[module] = true
+				modules = append(modules, module)
+			}
+		}
+
+	}
+
+	// Grab a more comprehensive list from the metricset keys, then reduce and merge
+	for mod := range r.metricSets {
+		// Skip over values we got from the modules map
+		if _, ok := dups[mod]; ok {
+			continue
+		}
+		modules = append(modules, mod)
+	}
```

Comment:
This can be a matter of style, but now that you are setting values in a map and appending to a list on the three sources, you could instead use a map to collect all the modules as the keys as if it were a set, and then at the end create the list of modules listing the keys of the map.

Not sure if I am explaining well, I mean to follow something like this pattern:
```
modules := make(map[string]bool)
for module := range r.modules {
    modules[module] = true
}
...
for module: = range sourceModules {
    modules[module] = true
}
...
for module: = range r.metricSets {
    modules[module] = true
}
modulesList := make([]string, 0, len(modules))
for module := range modules {
    modulesList = append(modulesList, module)
}
return modulesList
```

Reply:
No, that makes sense. Not a bad idea.